pkg/epp/datalayer/datasource.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package datalayer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+// DataSource is an interface required from all datalayer data collection
+// sources.
+type DataSource interface {
+	// Name returns the name of this datasource.
+	Name() string
+
+	// Start begins the collection process.
+	Start(ctx context.Context) error
+
+	// Stop stops the collection process.
+	Stop() error
+
+	// OutputType returns the type of information collected.
+	// TODO: this could be made private and excluded from the interface.
+	//       Use will likely be in the context of debugging/logging.
+	OutputType() reflect.Type
+
+	// AddExtractor adds an extractor to the data source.
+	// The extractor will be called whenever the data source might
+	// have some new raw information regarding an endpoint.
+	// The Extractor's expected input type is validated when it is
+	// registered.
+	AddExtractor(extractor Extractor) error
+
+	// TODO: the following is useful for a data source that operates on
+	// endpoints and might not be relevant for "global/system" collectors which
+	// might not need the concept of an endpoint. This can be split, if needed,
+	// to a separate interface in the future.
+
+	// AddEndpoint adds an endpoint to collect from.
+	AddEndpoint(ep Endpoint) error
+
+	// RemoveEndpoint removes an endpoint from collection.
+	RemoveEndpoint(ep Endpoint) error
```

Comment:
I see you wrote it yourself in the comments, it might make sense to have `DataSource` interface without these functions and additional `EndpointDataSource` interface that embeds `DataSource` and defines these functions.

Reply:
yes, that's a possibility.
As part of #1023 we agreed to push non-endpoint data source out until we have concrete requirements so I think we can keep it for now (i.e., don't have enough to go on at this time to rationalize what is shared and what's not between the per-endpoint and the non-endpoint data source).